FindMKL.cmake

```diff
@@ -0,0 +1,222 @@
+#
+#   https://cmake.org/cmake/help/v3.15/command/set.html#set-cache-entry
+#   https://cmake.org/cmake/help/v3.15/policy/CMP0077.html
+#
+if(NOT DEFINED MKL_ROOT)
+    set(MKL_ROOT $ENV{MKLROOT} CACHE PATH "MKL's root directory.")
+endif()
+
+if(NOT DEFINED MKL_PARALLEL)
+    set(MKL_PARALLEL ON CACHE BOOL "Toggles parallel and sequential versions.")
+endif()
+
+if(NOT DEFINED MKL_64BIT)
+    set(MKL_64BIT OFF CACHE BOOL "Toggles 32bit/64bit MKL interfaces.")
+endif()
+
+# Determine MKL's library folder
+#
+set(_mkl_libpath_suffix "lib/intel64")
+if(CMAKE_SIZEOF_VOID_P EQUAL 4) # 32 bit
+    set(_mkl_libpath_suffix "lib/ia32")
+endif()
+
+if (WIN32)
+    string(APPEND _mkl_libpath_suffix "_win")
+elseif (APPLE)
+    string(APPEND _mkl_libpath_suffix "_mac")
+else ()
+    string(APPEND _mkl_libpath_suffix "_lin")
+endif ()
+
+# Determine 32bit or 64bit interface is used
+#
+set(_mkl_lp "lp64")
+if(MKL_64BIT)
+    set(_mkl_lp "ilp64")
+endif()
+
+# Find MKL header
+#
+find_path(MKL_INCLUDE_DIR mkl.h
+    HINTS ${MKL_ROOT}/include
+    )
+mark_as_advanced(MKL_INCLUDE_DIR)
+
+# Find MKL core libraries
+#
+__mkl_find_library(MKL_CORE_LIB mkl_core)
+__mkl_find_library(MKL_INTERFACE_LIB mkl_intel_${_mkl_lp})
+
+if (MKL_PARALLEL)
+    if (CMAKE_CXX_COMPILER_ID STREQUAL "GNU" AND NOT APPLE)
+        __mkl_find_library(MKL_THREADING_LIB mkl_gnu_thread)
+    else ()
+        __mkl_find_library(MKL_THREADING_LIB mkl_intel_thread)
+    endif ()
+else ()
+    __mkl_find_library(MKL_THREADING_LIB mkl_sequential)
+endif ()
+
+# Find ScaLAPACK
+#
+set(MPI_DETERMINE_LIBRARY_VERSION ON)
+find_package(MPI)
+if (MPI_FOUND)
+    __mkl_find_library(MKL_SCALAPACK_LIB mkl_scalapack_${_mkl_lp})
+
+    string(FIND "${MPI_CXX_LIBRARY_VERSION_STRING}" "Open MPI" OMPI_POS)
+    if(OMPI_POS STREQUAL "-1")  # MPICH
+        if(APPLE)
+            __mkl_find_library(MKL_BLACS_LIB mkl_blacs_mpich_${_mkl_lp})
+        else()
+            __mkl_find_library(MKL_BLACS_LIB mkl_blacs_intelmpi_${_mkl_lp})
+        endif()
+    else()                      # OpenMPI
+        if(APPLE)
+            message(FATAL_ERROR "Only MPICH is supported on Apple.")
+        endif()
+         __mkl_find_library(MKL_BLACS_LIB mkl_blacs_openmpi_${_mkl_lp})
+    endif()
+
+    find_package_handle_standard_args(MKL_SCALAPACK REQUIRED_VARS MKL_BLACS_LIB
+                                                                  MKL_SCALAPACK_LIB)
+endif()
+
+find_package_handle_standard_args(MKL REQUIRED_VARS MKL_CORE_LIB 
+                                                    MKL_THREADING_LIB 
+                                                    MKL_INTERFACE_LIB 
+                                                    MKL_INCLUDE_DIR
+                                      HANDLE_COMPONENTS)
+
+
+# Define the MKL::MKL targets
+#
+if (MKL_FOUND AND NOT TARGET MKL::MKL)
+    find_package(Threads REQUIRED)
+
+    set(_mkl_threading_backend "")
+    if(MKL_PARALLEL)
+        find_package(OpenMP REQUIRED)
```

Comment:
`COMPOENTS CXX`?  and `find_dependency`

Reply:
Hmm, I was considering using `find_dependency` but did not commit since CMake's documentation suggested that it's used for `Config` files.

Reply:
Hmm, could question. I think you might be right, I am not sure about the reasoning, but I tend to agree

Reply:
I grepped over all Find modules shipped with CMake, none of them use `find_dependency`. I suggest we keep using `find_package` instead.